### PR TITLE
Scrape class attach metadata

### LINKED ISSUE #6752: ScrapeClasses: Support AttachMetadata
### Component(s)

Other (please comment)

### What is missing? Please describe.

I'd like to see Scrape Classes support configuring AttachMetadata.

I'm testing out making my prometheus deployments zone-aware.

I think this is possible with a scrape class that enables node metadata and a relabelling config to drop metrics targets from other zones.
Then I just need to create a Prometheus resource for each zone with a nodeSelector and the appropriate scrapeClass.

Having this on a default scrape class means that 
a) I don't have to go through and add it to every pod/service monitor I create
b) I don't have to hack it in to pod/service monitors I don't have full control over - e.g. deployed by 3rd party helm charts

### Describe alternatives you've considered.

A feature request to prometheus itself to expose a global config to enable node metadata by default.
A global config/flag in prometheus operator to enable node metadata by default, scrape classes seems a better fit though.

## Environment
Kubernetes Version: 1.30
Prometheus-Operator Version: 0.73

### PR DESCRIPTION
**Pull request by hamishforbes**
## Description
Fixes #6752

Adds support for configuring AttachMetadata via ScrapeClasses

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [x] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
Unit test

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add attachMetadata field to ScrapeClasses
```

### CODE REVIEW

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -868,11 +868,14 @@ func (cg *ConfigGenerator) generatePodMonitorConfig(
 	cfg = cg.AddTrackTimestampsStaleness(cfg, ep.TrackTimestampsStaleness)
 
 	var attachMetaConfig *attachMetadataConfig
-	if m.Spec.AttachMetadata != nil {
+	if m.Spec.AttachMetadata != nil || scrapeClass.AttachMetadata != nil {
 		attachMetaConfig = &attachMetadataConfig{
 			MinimumVersion: "2.35.0",
 			AttachMetadata: m.Spec.AttachMetadata,
 		}
+		if m.Spec.AttachMetadata == nil {
```
Comment (nicolastakashi) [inline pkg/prometheus/promcfg.go:876]: The configs defined on the monitor objects (PodMonitor, ServiceMonitor) are always taking precedence over scrape class.

It means the m.Spec.AttachMetadata will always have precedence over the scrapeClass.AttachMetadata.

Comment (nicolastakashi) [reply pkg/prometheus/promcfg.go:876]: Maybe worth create a function named GetAttachedMetadata and we implement this precedency logic inside this function and just reuse it, wdyt?

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 2]
```diff
@@ -1975,4 +1975,7 @@ type ScrapeClass struct {
 	//
 	// +optional
 	MetricRelabelings []RelabelConfig `json:"metricRelabelings,omitempty"`
+
+	// +optional
```
Comment (nicolastakashi) [inline pkg/apis/monitoring/v1/prometheus_types.go:1979]: Worth we point to the doc about AttachedMetadata as we're doing with relabel configs

**Comment by nicolastakashi** [comment]
Thanks @hamishforbes good progress 💪🏽 
Left some comments, but it's a good start

File: pkg/prometheus/promcfg.go (hunk 1 above)
Comment (hamishforbes) [reply pkg/prometheus/promcfg.go:876]: > The configs defined on the monitor objects (PodMonitor, ServiceMonitor) are always taking precedence over scrape class.
> 
> It means the m.Spec.AttachMetadata will always have precedence over the scrapeClass.AttachMetadata.

Yes, I thought that was the intention for Scrape classes
 As per the [proposal](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/proposals/202305-scrapeclasses.md#non-goals) its a non-goal to `Provide a way for Prometheus administrators to enforce/override settings on scrape configurations.`

> Maybe worth create a function named GetAttachedMetadata and we implement this precedency logic inside this function and just reuse it, wdyt?

Yeah that's probably cleaner, will do.

File: pkg/prometheus/promcfg.go [hunk 3]
```diff
@@ -433,6 +433,20 @@ func mergeTLSConfigWithScrapeClass(tlsConfig *monitoringv1.TLSConfig, scrapeClas
 	return tlsConfig
 }
 
+func mergeAttachMetadataWithScrapeClass(attachMetadata *monitoringv1.AttachMetadata, scrapeClass monitoringv1.ScrapeClass, minimumVersion string) *attachMetadataConfig {
```
Comment (nicolastakashi) [inline pkg/prometheus/promcfg.go:436]: wdyt
```suggestion
func mergeAttachMetadataWithScrapeClass(attachMetadata *monitoringv1.AttachMetadata, scrapeClass monitoringv1.ScrapeClass, minimumVersion string) *attachMetadataConfig {
	if attachMetadata == nil && scrapeClass.AttachMetadata == nil {
		return nil
	}
	if attachMetadata == nil {
		attachMetadata = scrapeClass.AttachMetadata
	}
	return &attachMetadataConfig{
		MinimumVersion: minimumVersion,
		AttachMetadata: attachMetadata,
	}
}
```

wdyt?

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -867,13 +881,7 @@ func (cg *ConfigGenerator) generatePodMonitorConfig(
...
 	cfg = cg.AddTrackTimestampsStaleness(cfg, ep.TrackTimestampsStaleness)
 
-	var attachMetaConfig *attachMetadataConfig
-	if m.Spec.AttachMetadata != nil {
-		attachMetaConfig = &attachMetadataConfig{
-			MinimumVersion: "2.35.0",
-			AttachMetadata: m.Spec.AttachMetadata,
-		}
-	}
+	var attachMetaConfig = mergeAttachMetadataWithScrapeClass(m.Spec.AttachMetadata, scrapeClass, "2.35.0")
```
Comment (nicolastakashi) [inline pkg/prometheus/promcfg.go:884]:
```suggestion
	attachMetaConfig := mergeAttachMetadataWithScrapeClass(m.Spec.AttachMetadata, scrapeClass, "2.35.0")
```

File: pkg/prometheus/promcfg.go [hunk 5]
```diff
@@ -1364,13 +1372,7 @@ func (cg *ConfigGenerator) generateServiceMonitorConfig(
...
 	}
 
-	var attachMetaConfig *attachMetadataConfig
-	if m.Spec.AttachMetadata != nil {
-		attachMetaConfig = &attachMetadataConfig{
-			MinimumVersion: "2.37.0",
-			AttachMetadata: m.Spec.AttachMetadata,
-		}
-	}
+	var attachMetaConfig = mergeAttachMetadataWithScrapeClass(m.Spec.AttachMetadata, scrapeClass, "2.37.0")
```
Comment (nicolastakashi) [inline pkg/prometheus/promcfg.go:1375]:
```suggestion
	attachMetaConfig := mergeAttachMetadataWithScrapeClass(m.Spec.AttachMetadata, scrapeClass, "2.37.0")
```

File: pkg/prometheus/promcfg.go (hunk 3 above)
Comment (hamishforbes) [reply pkg/prometheus/promcfg.go:436]: Yep, makes the logic clearer 👍

**Comment by hamishforbes** [comment]
@nicolastakashi Are you happy with the test coverage?
And do you want me to squash/rebase all the commits in here or do you guys prefer the full commit history?

**Review by nicolastakashi — APPROVED** [review]

**Pull request merged**